### DISCUSSION TITLE
I love this project!

### DISCUSSION THREAD

**Discussion by Montclair** [wqweto/VbAsyncSocket#24] [General]
I just wanted to say thank you for coding this.  It's truly fantastic and has freed my older systems from TLS issues everywhere on the net.  I'm using it to connect everywhere without issue.  For example, I decided today that I needed one of my programs to be able to send email.  All the stuff I found online was dated and / or didn't work properly older systems.  Your simple example in the readme put me on the right track and with some research into SMTP commands and responses, I've got my program sending test emails, and am nearly finish coding a fully functional routine.

I wouldn't have been able to do any of this without paying for some third party DLL, considering the older system this program runs on.  And, being able to make a DLL without changing a line of code (just had to change one property on three classes) opens up this same access to everything for my older VBA applications.

_**Thank you so much for this fantastic project!**_

**Comment by wqweto** [comment]
Great!

> (just had to change one property on three classes)

Which property? Can we fix this in trunk branch?

**Comment by Montclair** [comment]
"Instancing" had to be changed to "5 - MultiUse" for the three class modules, cAsyncSocket, cHttpRequest, and cTlsSocket.  That's only necessary to get the project to compile to a DLL.

**Reply by wqweto** [reply]
I see. Yes, this can only be private for Std-EXE projects so can't be enhanced in trunk.

**Reply by Montclair** [reply]
Right, and it's not a big deal for anyone to make that change if they want a DLL.

**Reply by wqweto** [reply]
Btw, you can try adding `ASYNCSOCKET_NO_TLSSERVER = 1` in project properties "Conditional Compilation" settings to reduce compiled DLL size.

This strips server-side TLS implementation which is not used in `cHttpRequest` in principle (still might be used standalone in `cTlsSocket`).

With server-side TLS is stripped then `cTlsSocket` raises run-time error on `InitServerTls` or any other server-side method being called.

**Reply by Montclair** [reply]
Great tip!  That cut the DLL size down by 25K and works for my purposes.  Thanks!

**Reply by etc6849** [reply]
Is the DLL something I can manually register and use a replacement winhttp object to be created and used in a vbs file or even from vbscript?  If so, would you be kind enough to share it?  I'm playing around with some XP VMs, and found out I can send messages using a telegram bot like I can under Windows 10.

**Reply by wqweto** [reply]
@etc6849 FYI, I just published a compiled ActiveX DLL with `cHttpRequest` replacement class under [releases](https://github.com/wqweto/VbAsyncSocket/releases/tag/HRR-1.0) in `HttpRequest.zip` archive under `Assets`.

This VBScript
```
With CreateObject("HttpRequest.cHttpRequest")
    .Open_ "GET", "https://www.howsmyssl.com/a/check"
    .Send
    WScript.echo Replace(.ResponseText, ",", vbCrLf)
End With
```
. . . can be executed like this:
```
C:> C:\Windows\SysWow64\cscript /nologo MyScript.vbs
{"given_cipher_suites":["TLS_AES_128_GCM_SHA256"
"TLS_AES_256_GCM_SHA384"
"TLS_CHACHA20_POLY1305_SHA256"
"TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256"
"TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256"
"TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384"
"TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384"
"TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256"
"TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256"
"TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA"
"TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA"
"TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA"
"TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA"
"TLS_RSA_WITH_AES_128_GCM_SHA256"
"TLS_RSA_WITH_AES_256_GCM_SHA384"
"TLS_RSA_WITH_AES_128_CBC_SHA"
"TLS_RSA_WITH_AES_256_CBC_SHA"]
"ephemeral_keys_supported":true
"session_ticket_supported":true
"tls_compression_supported":false
"unknown_cipher_suite_supported":false
"beast_vuln":false
"able_to_detect_n_minus_one_splitting":false
"insecure_cipher_suites":{}
"tls_version":"TLS 1.3"
"rating":"Probably Okay"}
```

**Reply by etc6849** [reply]
Wow.  Great work and very nice!  Works great on a 32 bit XP SP3 VM that I had upgraded ie 8 using the POSReady 2009 registry hack to enable TLS 1.1 and 1.2 for ie 8 as shown below.  However, having a certificate issue on an XP SP2 x64 VM as shown at the bottom of this post.

Microsoft Windows XP [Version 5.1.2600]
(C) Copyright 1985-2001 Microsoft Corp.

C:\Documents and Settings\etc>regsvr32

C:\Documents and Settings\etc>regsvr32 HttpRequest.dll

C:\Documents and Settings\etc>cscript "C:\Documents and Settings\etc\Desktop\Tes
t.vbs"
Microsoft (R) Windows Script Host Version 5.7
Copyright (C) Microsoft Corporation. All rights reserved.

{"given_cipher_suites":["TLS_AES_128_GCM_SHA256"
"TLS_AES_256_GCM_SHA384"
"TLS_CHACHA20_POLY1305_SHA256"
"TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256"
"TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256"
"TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384"
"TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384"
"TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256"
"TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256"
"TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA"
"TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA"
"TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA"
"TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA"
"TLS_RSA_WITH_AES_128_GCM_SHA256"
"TLS_RSA_WITH_AES_256_GCM_SHA384"
"TLS_RSA_WITH_AES_128_CBC_SHA"
"TLS_RSA_WITH_AES_256_CBC_SHA"]
"ephemeral_keys_supported":true
"session_ticket_supported":true
"tls_compression_supported":false
"unknown_cipher_suite_supported":false
"beast_vuln":false
"able_to_detect_n_minus_one_splitting":false
"insecure_cipher_suites":{}
"tls_version":"TLS 1.3"
"rating":"Probably Okay"}

However, on an XP SP2 x64 VM, it is giving the certificate issue below:

Microsoft Windows [Version 5.2.3790]
(C) Copyright 1985-2003 Microsoft Corp.

C:\>%SystemRoot%\Syswow64\regsvr32 HttpRequest.dll

C:\>C:\Windows\SysWow64\cscript /nologo "C:\Documents and Settings\Administrato\Desktop\test.vbs"

cTlsSocket.pvPkiCertValidate: The certificate or certificate chain is based on an untrusted root

**Reply by wqweto** [reply]
Yes, this error is correct to get if you think about it. There is no valid root certificate installed on your XP SP2 for the site you are accessing chain of certificates to be validated to a valid trust anchor.

So you have to either install a valid root certificate like [`ISRG_Root_X1.cer`](https://github.com/wqweto/VbAsyncSocket/blob/master/lib/ISRG_Root_X1.cer) from repo's `lib` dirtectory (being Let's Encrypt root certificate) or use WinHttpRequest's `Option` property to disable checking of root certificate validity like this
```
Const WinHttpRequestOption_SslErrorIgnoreFlags = 4
Const SslErrorFlag_UnknownCA = 256

With CreateObject("HttpRequest.cHttpRequest")
    .Option_(WinHttpRequestOption_SslErrorIgnoreFlags) = SslErrorFlag_UnknownCA
    .Open_ "GET", "https://www.howsmyssl.com/a/check"
    .Send
    WScript.echo Replace(.ResponseText, ",", vbCrLf)
End With
```
Edit: Actually there is a 3-rd option -- to pre-load [`ca-bundle.pem`](https://github.com/wqweto/VbAsyncSocket/blob/master/test/Basic/ca-bundle.pem) file which contains all modern root CA certificates as a custom `WinHttpRequestOption_RootCA` option but I'll have to recompile the DLL to support this.

**Reply by etc6849** [reply]
Thanks!  Installing the root certificate fixed the error, disregard my other post that I deleted :)

Do you have a way we can donate to you?  This DLL fixed a huge problem with my custom home automation setup, and I want to thank you.

**Reply by etc6849** [reply]
All the GETs now work on my HA server.  However, I'm having trouble sending an image which was working under Windows 10 using MSXML2.XMLHTTP.

What the codes does is reads in an image stored on the hard drive (originally captured from a webcam that gets triggered when the doorbell is rang).   Then it sends the image using an HTTP Post.

As before, this is probably due to my lack of understanding something, but of course any help or hints are always much appreciated :)

EDIT: oXMLHTTP.Send bytePayLoad is generating the error:
The operation timed out (-2147012894)

--------------------------------------------------------------
WinHttpRequest.vbs (works in Windows 7):
--------------------------------------------------------------
```
Const WinHttpRequestOption_SslErrorIgnoreFlags = 4
Const SslErrorFlag_UnknownCA = 256
Const URL = "https://api.telegram.org/bot"
Const TOKEN = "xxx"
Const METHOD_NAME = "/sendPhoto?"
Const CHAT_ID = "xxx"
Const FOLDER = "C:\"
Const JPG_FILE = "Image1.jpg"

Set data = CreateObject("Scripting.Dictionary")
data.Add "chat_id", CHAT_ID

' generate boundary
For n = 1 To 16: s = s & Chr(65 + Int(Rnd * 25)): Next
BOUNDARY = s & CDbl(Now)

For Each key In data.keys
    part = part & "--" & BOUNDARY & vbCrLf
    part = part & "Content-Disposition: form-data; name=""" & key & """" & vbCrLf & vbCrLf
    part = part & data(key) & vbCrLf
Next

' filename
part = part & "--" & BOUNDARY & vbCrLf
part = part & "Content-Disposition: form-data; name=""photo"";filename=""" & JPG_FILE & """" & vbCrLf & vbCrLf

' read jpg file as binary
Set ado = CreateObject("ADODB.Stream")
ado.Type = 1 'binary
ado.Open
ado.LoadFromFile FOLDER & JPG_FILE
ado.Position = 0
jpg = ado.Read
ado.Close

' combine part, jpg , end
ado.Open
ado.Position = 0
ado.Type = 1 ' binary
ado.Write ToBytes(part)
ado.Write jpg
ado.Write ToBytes(vbCrLf & "--" & BOUNDARY & "--")
ado.Position = 0

reqURL = URL & TOKEN & METHOD_NAME
With CreateObject("WinHttp.WinHttpRequest.5.1")
    .Open "POST", reqURL, False
    .setRequestHeader "Content-Type", "multipart/form-data;boundary=" & BOUNDARY
    .Option(WinHttpRequestOption_SslErrorIgnoreFlags) = SslErrorFlag_UnknownCA
    .send ado.Read
    WScript.echo Replace(.ResponseText, ",", vbCrLf)
End With

Function ToBytes(stuff)
    With CreateObject("ADODB.Stream")
       .Open
       .Type = 2 ' text
       .Charset = "_autodetect"
       .WriteText stuff
       .Position = 0
       .Type = 1
       ToBytes= .Read
       .Close
    End With
End Function
```

--------------------------------------------------------------
cHttpRequest.vbs (doesn't work on windows 7 or any XP version I have):
--------------------------------------------------------------
```
Const WinHttpRequestOption_SslErrorIgnoreFlags = 4
Const SslErrorFlag_UnknownCA = 256
Const URL = "https://api.telegram.org/bot"
Const TOKEN = "xxx"
Const METHOD_NAME = "/sendPhoto?"
Const CHAT_ID = "xxx"
Const FOLDER = "C:\"
Const JPG_FILE = "Image1.jpg"

Set data = CreateObject("Scripting.Dictionary")
data.Add "chat_id", CHAT_ID

' generate boundary
For n = 1 To 16: s = s & Chr(65 + Int(Rnd * 25)): Next
BOUNDARY = s & CDbl(Now)

For Each key In data.keys
    part = part & "--" & BOUNDARY & vbCrLf
    part = part & "Content-Disposition: form-data; name=""" & key & """" & vbCrLf & vbCrLf
    part = part & data(key) & vbCrLf
Next

' filename
part = part & "--" & BOUNDARY & vbCrLf
part = part & "Content-Disposition: form-data; name=""photo"";filename=""" & JPG_FILE & """" & vbCrLf & vbCrLf

' read jpg file as binary
Set ado = CreateObject("ADODB.Stream")
ado.Type = 1 'binary
ado.Open
ado.LoadFromFile FOLDER & JPG_FILE
ado.Position = 0
jpg = ado.Read
ado.Close

' combine part, jpg , end
ado.Open
ado.Position = 0
ado.Type = 1 ' binary
ado.Write ToBytes(part)
ado.Write jpg
ado.Write ToBytes(vbCrLf & "--" & BOUNDARY & "--")
ado.Position = 0

reqURL = URL & TOKEN & METHOD_NAME
With CreateObject("WinHttp.WinHttpRequest.5.1")
    .Open "POST", reqURL, False
    .setRequestHeader "Content-Type", "multipart/form-data;boundary=" & BOUNDARY
    .Option(WinHttpRequestOption_SslErrorIgnoreFlags) = SslErrorFlag_UnknownCA
    .send ado.Read
    WScript.echo Replace(.ResponseText, ",", vbCrLf)
End With

Function ToBytes(stuff)
    With CreateObject("ADODB.Stream")
       .Open
       .Type = 2 ' text
       .Charset = "_autodetect"
       .WriteText stuff
       .Position = 0
       .Type = 1
       ToBytes= .Read
       .Close
    End With
End Function

```

**Reply by wqweto** [reply]
Yes, there was a problem sending byte-arrays on POST which is fixed in 62114beac9966324b63f13d6524f8f9637bb308a and included in [1.0.1 release](https://github.com/wqweto/VbAsyncSocket/releases/tag/HRR-1.0.1) of the binaries. Let me know if this fixes upload with your code above.

Latest version also allows loading custom CAs from a PEM file like this
```
Const WinHttpRequestOption_RootCA = 20
...
    .Option_(WinHttpRequestOption_RootCA) = "C:\path\to\ca-bundle.pem"
```
This is compatible with **curl** project [CA certificates extracted from Mozilla](https://curl.se/docs/caextract.html) bundle file they maintain for their own purposes.

Edit: Btw, with stock `WinHttpRequest` and here with `cHttpRequest` it is more memory efficient to directly send the `ADODB.Stream` object (or any other `IStream` interface implementing object) without first reading its content into a byte-array i.e. directly calling `.Send ado` instead of `.Send ado.Read`.

In this case you don't need to reposition the stream at the beginning beforehand too i.e. setting `ado.Position = 0` is not needed as no `ado.Read` is called.

**Reply by etc6849** [reply]
EDIT: it works now.  I forgot how I was calling it in Windows x64 and used the 64 bit cScript.exe instead of the 32 bit cScript.exe.

Many thanks for the scripting hint!  Will take a look.  I mainly learned programming (really just scripting) for my home automation hobby, but I'm always interested in learning and improving.  Not going to quit my day job (has nothing to do with programming) yet though.

I de-registered the old version and deleted the old dll, and the new dll registers ok.  However, windows will not let me use it and gives the error: Activex component cna't create object 'HttpRequest.cHttpRequest'.

**Reply by etc6849** [reply]
Please see my edit above.  Works great.  Thanks!

**Comment by wqweto** [comment]
Just to let you know that in commit 41e1ef3ef10a930e2dea74fb0a930e3471d34dee a latent [heisenbug](https://en.wikipedia.org/wiki/Heisenbug) was fixed which prevented compiled `cTlsSocket` instances from establishing connection to some servers while working ok in VB6 IDE.

The bug stemmed from a bug in the VB6 compiler and affected native compiled binaries only.

Update your sources to latest revision if you can and want to.
